src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,154 @@
 package core.basesyntax;
 
+import java.util.NoSuchElementException;
+
 public class ArrayList<T> implements List<T> {
+    private static final int DEFAULT_SIZE = 10;
+    private static final int NOT_FOUND_INDEX = -1;
+    private T[] content;
+    private int size;
+
+    public ArrayList() {
+        content = (T[]) new Object[DEFAULT_SIZE];
+    }
+
+    public ArrayList(int customSize) {
+        content = (T[]) new Object[customSize];
+    }
+
     @Override
     public void add(T value) {
-
+        growIfNeeded();
+        content[size++] = value;
     }
 
     @Override
     public void add(T value, int index) {
-
+        if (!(index >= 0 && index <= size)) {
+            String errorMessage = String.format("Can't add value at index %s. It's out of bounds.",
+                    index);
+            throw new ArrayListIndexOutOfBoundsException(errorMessage);
+        }
+        growIfNeeded();
+        System.arraycopy(content, index, content, index + 1, size++ - index);
+        content[index] = value;
     }
 
     @Override
     public void addAll(List<T> list) {
+        int listSize = list.size();
+        growIfNeeded(listSize);
 
+        T[] listArray = convertListToArray(list);
+        System.arraycopy(listArray, 0, content, size, listSize);
+        size += listSize;
     }
 
     @Override
     public T get(int index) {
-        return null;
+        if (isInRange(index)) {
+            return content[index];
+        }
+        String errorMessage = String.format("Can't get value with index %s. It's out of bounds.",
+                index);
+        throw new ArrayListIndexOutOfBoundsException(errorMessage);
     }
 
     @Override
     public void set(T value, int index) {
-
+        if (!isInRange(index)) {
+            String errorMessage = String.format("Can't set value at index %s. It's out of bounds.",
+                    index);
+            throw new ArrayListIndexOutOfBoundsException(errorMessage);
+        }
```

Comment:
Also, let's extract to a separate method.
```suggestion
        verifyIndexIsInRange();
```